group-coordinator/src/main/java/org/apache/kafka/coordinator/group/ShareGroupAutoOffsetResetStrategy.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group;
+
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.common.config.ConfigException;
+import org.apache.kafka.common.requests.ListOffsetsRequest;
+import org.apache.kafka.common.utils.Utils;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.Optional;
+
+public class ShareGroupAutoOffsetResetStrategy {
```

Comment:
As mention in https://github.com/apache/kafka/pull/17573#discussion_r1814871532, let each `AutoOffsetResetStrategy` evolve themselves

Reply:
Can we have javadocs please on the class.

Reply:
Moreover the purpose of having a separate enum as `ShareGroupAutoOffsetReset` was to just mention that ShareGroup do no support all offset reset strategy but the ones explicitly mentioned in the `ShareGroupAutoOffsetReset`. Which is the sub strategies from `AutoOffsetResetStrategy`.

This class seems same as `AutoOffsetResetStartegy`, why do we need a different one? If we do need one then do we need to copy the methods as it is or could just use  `ShareGroupAutoOffsetResetStrategy` as a wrapper over `AutoOffsetResetStrategy` to differentiate with self evolving startegies of ShareGroup.

Reply:
I think we still need a separate `ShareGroupAutoOffsetResetStrategy`, because this allows us to separate the enum and validation logic. As for the wrapper part you mentioned, perhaps we could use the Delegation pattern to reuse `AutoOffsetResetStrategy`'s logic. What do you think?

Reply:
> I think we still need a separate ShareGroupAutoOffsetResetStrategy

Yes, that's fine. 

> perhaps we could use the Delegation pattern

That could be a good option.

Reply:
I wouldn't over-engineer this. It's a pretty trivial class and there's not much saved. Up to you in this case from my point of view.

Reply:
@apoorvmittal10 Would like to know your opinion on this.